lib/dbutils.js

```diff
@@ -0,0 +1,832 @@
+                            return 0;
+                        } else if (val1 < val2) {
+                            return -1;
+                        } else {
+                            return 1;
+                        }
+                    }
+                });
+            }
+            return null;
+        },
+        type: 'blob'
+    };
+}
+
+function generateConverters(schema) {
+    schema.converters = {};
+    Object.keys(schema.attributes).forEach(function(key) {
+        var set_type = /^set<(\w+)>$/.exec(schema.attributes[key]);
+        var obj_type = set_type ? set_type[1] : schema.attributes[key];
+        var obj_converter = dbu.conversions[obj_type] || {};
+        if (!obj_converter.write) {
+            obj_converter.write = function(val) {
+                return val;
+            };
+        }
+        if (!obj_converter.read) {
+            obj_converter.read = function(val) {
+                return val;
+            };
+        }
+        if (!obj_converter.type) {
+            obj_converter.type = obj_type;
+        }
+        if (set_type) {
+            schema.converters[schema.attributes[key]] = generateSetConverter(obj_converter);
+        } else {
+            schema.converters[schema.attributes[key]] = obj_converter;
+        }
+    });
+    return schema;
+}
+
+/*
+ * Error instance wrapping HTTP error responses
+ *
+ * Has the same properties as the original response.
+ */
+function HTTPError(response) {
+    Error.call(this);
+    Error.captureStackTrace(this, HTTPError);
+    this.name = this.constructor.name;
+    this.message = JSON.stringify(response);
+
+    for (var key in response) {
+        this[key] = response[key];
+    }
+}
+
+dbu.HTTPError = HTTPError;
+
+dbu.hashKey = function hashKey(key) {
+    return crypto.Hash('sha1')
+    .update(key)
+    .digest()
+    .toString('base64')
+        // Replace [+/] from base64 with _ (illegal in Cassandra)
+    .replace(/[+\/]/g, '_')
+        // Remove base64 padding, has no entropy
+    .replace(/=+$/, '');
+};
+
+dbu.getValidPrefix = function getValidPrefix(key) {
+    var prefixMatch = /^[a-zA-Z0-9_]+/.exec(key);
+    if (prefixMatch) {
+        return prefixMatch[0];
+    } else {
+        return '';
+    }
+};
+
+dbu.makeValidKey = function makeValidKey(key, length) {
+    var origKey = key;
+    key = key.replace(/_/g, '__')
+    .replace(/\./g, '_');
+    if (!/^[a-zA-Z0-9_]+$/.test(key)) {
+        // Create a new 28 char prefix
+        var validPrefix = dbu.getValidPrefix(key).substr(0, length * 2 / 3);
+        return validPrefix + dbu.hashKey(origKey).substr(0, length - validPrefix.length);
+    } else if (key.length > length) {
+        return key.substr(0, length * 2 / 3) + dbu.hashKey(origKey).substr(0, length / 3);
+    } else {
+        return key;
+    }
+};
+
+dbu.keyspaceName = function(domain, bucket) {
+    var prefix = dbu.makeValidKey(domain, Math.max(26, 48 - bucket.length - 3));
```

Comment:
This could perhaps be simpler in SQLite: http://stackoverflow.com/questions/3694276/what-are-valid-table-names-in-sqlite

Basically, with brackets we don't really need to do much escaping / hashing at all.